### ISSUE TITLE
Docker for windows permission check fails on /app volume

### ISSUE THREAD

**Issue by kvij** [lando/lando#705]
On OSX and Linux mount the /app volume is mounted with file permissions and ownership as they should be. On windows the ownership and group is always root. The file permissions are always 755. The user www-data can for some odd reason still access(write) everything but as soon as the application actually checks the permissions before taking action it fails. chown and chmod have no effect.

Possible solutions:

- Force the mount to www-data (at least on appserver) if at all possible
- Integrate a service like http://docker-sync.io/

The added benefit to a sync solution is of course the performance upgrade on Windows and OSX.

**Comment by Shurrir** [comment]
+2 here

Running a Drupal site, for some weird reason when I first upload a file the files folder is created, but after that it's impossible to upload another file (no permission)
Tried changing the folder permissions in every way I could think of, but nothing seems to work.

**Comment by kvij** [comment]
@Shurrir That is exactly how we ran into this problem as well. The file field uses core/includes/file.inc file_prepare_directory() in which the first time the directory is created and when that succeeds the file is uploaded. If the directory already exists is_writable() is called what returns false. Unfortunately there are a lot of calls to is_writable() in Drupal.

**Comment by dustinleblanc** [comment]
What version of Lando and Windows is affected here? We'll have to be able to reliably reproduce to get this into a sprint to fix. If you could provide your Windows version, Docker version, and Lando version @kvij  and @Shurrir that will help us narrow it down.

**Comment by Jesperanza** [comment]
I may be experiencing a similar problem. I am on Windows 10 Pro, Lando v3.0.0-beta.35, and Docker Version 17.12.0-ce-win47 (15139). I pulled my Drupal 7 site to local environment using the Pantheon recipe and noticed in my status reports multiple errors relating to permissions and  directories. As far as I was able to tell, my directories are configured correctly, but there does seem to be a problem with Drupal being allowed to write and create/add files/folders. I also am getting an error when i try to upload a file. Partial errrors are listed below. 

"..could not be created due to a misconfigured files directory. Please ensure that the files directory is correctly configured and that the webserver has permission to create directories."

"...directory is not writable."

"The file could not be uploaded."

**Comment by Shurrir** [comment]
Windows 10 Pro
Version 1709
Was running the latest Lando & Docker versions since it was a fresh install.
Same for my colleague, who had the same issue.

**Comment by blakestein** [comment]
I ran into this issue with Docksal as well, I'm not sure there's a fix at the moment. What I ended up doing was creating a files directory in my user HOME dir and symlinking it to sites/default/files. I didn't think that was ideal since you'd lose your files unless you also back them up, so I'm back to using a VM :( 

https://github.com/docker/for-win/issues/497

**Comment by pirog** [comment]
Yeah unfortunately this is a known (and fairly widespread) issue that effects people using drupal and docker for windows. 

Docker for Windows uses SAMBA for file sharing which sets all dir/files shared into docker as `777`. 

We've experimented with other file sharing approaches in our previous Kalabox project but they all come with their own set of issues. Plus i don't think we want to get into a situation where we have to ALSO handle our own file sharing system, potentially for each OS. Trying to maintain and keep a PaaS for your computer (which is more or less what lando is) is tough enough. 

So tl;dr here is i dont think its very desirable for us to do anything until this is fixed upstream in either docker or drupal.

**Comment by kvij** [comment]
@pirog I thought that docker-sync would solve those problems but I see that it requires rather a rather elaborate setup with the _Windows Subsystem for Linux_ on Windows. Therefore I can do nothing but agree with Lando not being the place to solve this. I do hope we can add a documented solution for this problem

**Comment by pirog** [comment]
@kvij agreed.

i assumed you've looked at the core patch (or equivalent) over yonder
https://docs.devwithlando.io/issues/drupal-win-upload.html ?

**Comment by blakestein** [comment]
@pirog thanks for pointing that out, the latest patch from that issue seems to fix this issue for me 

https://www.drupal.org/node/944582
